### PR TITLE
Simplify Rule

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes #.

## Rationale for this change
Updates `OptimizeAggregateOrder` rule for better extensibility in the future. As pointed out in the [discussion](https://github.com/apache/arrow-datafusion/pull/10061#discussion_r1564030050). Current implementation overfits on the `FIRST_VALUE`, `LAST_VALUE`. This PR extends support to other aggregators given they implement necessary `API`s.

I also updated comments to communicate better what is being optimized by the rule.

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/update_aggr_exprs.rs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+/// Updates the aggregators with mode `AggregateOrderSensitivity::Beneficial`
+/// if existing ordering enables to execute them more efficiently.
+fn update_aggregator_when_beneficial(
+    plan: Arc<dyn ExecutionPlan>,
+) -> Result<Transformed<Arc<dyn ExecutionPlan>>> {
+    if let Some(aggr_exec) = plan.as_any().downcast_ref::<AggregateExec>() {
+        if !aggr_exec.mode().is_first_stage() {
+            return Ok(Transformed::no(plan));
+        }
```
Comment (berkaysynnada) [inline datafusion/core/src/physical_optimizer/update_aggr_exprs.rs]: Why does this rule work on only first stage? A comment would be helpful.

File: datafusion/core/src/physical_optimizer/update_aggr_exprs.rs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        let input = aggr_exec.input();
+        let mut aggr_expr = aggr_exec.aggr_expr().to_vec();
+        let group_by = aggr_exec.group_by();
+
+        let input_eq_properties = input.equivalence_properties();
+        let groupby_exprs = group_by.input_exprs();
+        // If existing ordering satisfies a prefix of the GROUP BY expressions,
+        // prefix requirements with this section. In this case, aggregation will
+        // work more efficiently.
+        let indices = get_ordered_partition_by_indices(&groupby_exprs, input);
```
Comment (berkaysynnada) [inline datafusion/core/src/physical_optimizer/update_aggr_exprs.rs]: `get_ordered_partition_by_indices()` is used here for group_by exprs. Perhaps the function name should be generalized.

File: datafusion/core/src/physical_optimizer/update_aggr_exprs.rs (hunk 1 above)
Comment (mustafasrepo) [reply datafusion/core/src/physical_optimizer/update_aggr_exprs.rs]: At final stages, `merge_batch` is used. `merge_batch` doesn't assume any order regarding its input (its input may lose required ordering fields, during `PartialAggregation`. Necessary information for proper merge is carried in the partial state).

**Comment by ozankabak** [comment]
The final version of this work has been merged upstream.

**Pull request closed without merging**